boba-chain-ops/ether/migrate.go

```diff
@@ -0,0 +1,299 @@
+				"account has non-zero balance in state - should never happen",
+				"addr", addr,
+				"balance", bal.String(),
+			)
+			if !noCheck {
+				errCh <- fmt.Errorf("account has non-zero balance in state - should never happen: %s", addr.String())
+				return
+			}
+		}
+
+		// Add balances to the total found.
+		switch slotType {
+		case BalanceSlot:
+			// Send the data to the channel.
+			outCh <- accountData{
+				balance:    value.Big(),
+				legacySlot: key,
+				address:    addr,
+			}
+		case AllowanceSlot:
+			// Allowance slot. Do nothing here.
+		default:
+			// Should never happen.
+			if noCheck {
+				log.Error("unknown slot type", "slot", key, "type", slotType)
+				errCh <- nil
+			} else {
+				log.Crit("unknown slot type %d, should never happen", slotType)
+			}
+		}
+
+		errCh <- nil
+		return
+	})
+
+	if err != nil {
+		return err
+	}
+
+	// Close the outCh to cancel the collector. The collector will signal that it's done
+	// using doneCh. Any values waiting to be read from outCh will be read before the
+	// collector exits.
+	close(outCh)
+	<-doneCh
+
+	// Log how many slots were iterated over.
+	log.Info("Iterated legacy balances", "count", count)
+
+	// Verify the supply delta. Recorded total supply in the LegacyERC20ETH contract may be higher
+	// than the actual migrated amount because self-destructs will remove ETH supply in a way that
+	// cannot be reflected in the contract. This is fine because self-destructs just mean the L2 is
+	// actually *overcollateralized* by some tiny amount.
+	totalSupply := g.Alloc[predeploys.LegacyERC20ETHAddr].Storage[CalcOVMETHTotalSupplyKey()].Big()
+	delta := new(big.Int).Sub(totalSupply, totalFound)
+	if delta.Cmp(expDiff) != 0 {
+		log.Error(
+			"supply mismatch",
+			"migrated", totalFound.String(),
+			"supply", totalSupply.String(),
+			"delta", delta.String(),
+			"exp_delta", expDiff.String(),
+		)
+		if !noCheck {
+			return fmt.Errorf("supply mismatch: %s", delta.String())
+		}
+	}
+
+	// Supply is verified.
+	log.Info(
+		"supply verified OK",
+		"migrated", totalFound.String(),
+		"supply", totalSupply.String(),
+		"delta", delta.String(),
+		"exp_delta", expDiff.String(),
+	)
+
+	// Set the total supply to 0. We do this because the total supply is necessarily going to be
+	// different than the sum of all balances since we no longer track balances inside the contract
+	// itself. The total supply is going to be weird no matter what, might as well set it to zero
+	// so it's explicitly weird instead of implicitly weird.
+	SetTotalSupply(g)
+	log.Info("Set the totalSupply to 0")
+
+	return nil
+}
+
+func IterateState(g *types.Genesis, cb func(key, value common.Hash, errCh chan error)) error {
+	// Deep copy
+	storage := make(map[common.Hash]common.Hash)
+	for key, value := range g.Alloc[predeploys.LegacyERC20ETHAddr].Storage {
+		storage[key] = value
+	}
+
+	errCh := make(chan error, len(storage))
+
+	for key, value := range storage {
+		go cb(key, value, errCh)
```

Comment:
Nothing strictly wrong with this approach, but, I think it would be cleaner if we were to eliminate the errCh construct and simply use the sync/errgroup package.  We could leave the original return pattern of directly returning errors and not dealing with the channel logic.

Something like:

```
var eg errgroup.Group
for key, value := range storage {
    eg.Go(cb)
}
if err := eg.Wait(); err != nil {
    return errr
}
```

Reply:
Yes, the channel is removed and replaced by `errgroup`